src/materials/DamagePlasticityStressUpdate.C

```diff
@@ -0,0 +1,680 @@
+                        "severe hardening/softening this may be "
+                        "suboptimal.");
+  params.addClassDescription("Damage Plasticity Model for concrete");
+  return params;
+}
+
+DamagePlasticityStressUpdate::DamagePlasticityStressUpdate(const InputParameters & parameters)
+  : MultiParameterPlasticityStressUpdate(parameters, 3, 1, 2),
+    _f_tol(getParam<Real>("yield_function_tol")),
+
+    _alfa(getParam<Real>("biaxial_uniaxial_compressive_stress_factor")),
+    _alfa_p(getParam<Real>("dilatancy_factor")),
+    _s0(getParam<Real>("stiff_recovery_factor")),
+
+    _Chi(getParam<Real>("ft_ep_slope_factor_at_zero_ep")),
+    _Dt(getParam<Real>("tensile_damage_at_half_tensile_strength")),
+    _ft(getParam<Real>("yield_strength_in_tension")),
+    _FEt(getParam<Real>("fracture_energy_in_tension")),
+
+    _fyc(getParam<Real>("yield_strength_in_compression")),
+    _Dc(getParam<Real>("compressive_damage_at_max_compressive_strength")),
+    _fc(getParam<Real>("maximum_strength_in_compression")),
+    _FEc(getParam<Real>("fracture_energy_in_compression")),
+
+    _at(1.5 * std::sqrt(1 - _Chi) - 0.5),
+    _ac((2. * (_fc / _fyc) - 1. + 2. * std::sqrt(std::pow((_fc / _fyc), 2.) - _fc / _fyc))),
+
+    _zt((1. + _at) / _at),
+    _zc((1. + _ac) / _ac),
+    _dPhit(_at * (2. + _at)),
+    _dPhic(_ac * (2. + _ac)),
+    _sqrtPhit_max((1. + _at + std::sqrt(1. + _at * _at)) / 2.),
+    _sqrtPhic_max((1. + _ac) / 2.),
+    _dt_bt(log(1. - _Dt) / log((1. + _at - std::sqrt(1. + _at * _at)) / (2. * _at))),
+    _dc_bc(log(1. - _Dc) / log((1. + _ac) / (2. * _ac))),
+    _ft0(0.5 * _ft /
+         ((1. - _Dt) * std::pow((_zt - _sqrtPhit_max / _at), (1. - _dt_bt)) * _sqrtPhit_max)),
+    _fc0(_fc / ((1. - _Dc) * std::pow((_zc - _sqrtPhic_max / _ac), (1. - _dc_bc)) * _sqrtPhic_max)),
+    _small_smoother2(Utility::pow<2>(getParam<Real>("tip_smoother"))),
+ 
+
+    _sqrt3(std::sqrt(3.)),
+    _perfect_guess(getParam<bool>("perfect_guess")),
+    _eigvecs(RankTwoTensor()),
+    _intnl0(declareProperty<Real>("damage_state_in_tension")),
+    _intnl1(declareProperty<Real>("damage_state_in_compression")),
+    _ele_len(declareProperty<Real>("element_length")),
+    _gt(declareProperty<Real>("fracture_energy_in_tension")),
+    _gc(declareProperty<Real>("fracture_energy_in_compression")),
+    _tD(declareProperty<Real>("tensile_damage")),
+    _cD(declareProperty<Real>("compression_damage")),
+    _D(declareProperty<Real>("damage_variable")),
+    _min_ep(declareProperty<Real>("min_ep")),
+    _mid_ep(declareProperty<Real>("mid_ep")),
+    _max_ep(declareProperty<Real>("max_ep")),
+    _sigma0(declareProperty<Real>("damaged_min_principal_stress")),
+    _sigma1(declareProperty<Real>("damaged_mid_principal_stress")),
+    _sigma2(declareProperty<Real>("damaged_max_principal_stress"))
+{
+}
+
+void
+DamagePlasticityStressUpdate::initQpStatefulProperties()
+{
+  // if (_current_elem->n_vertices() < 3)
```

Comment:
Put a comment saying that there multiple potential ways to determine the correct element length, and these are some options that are still being considered.

Reply:
Addressed